src/misc/lv_array.h

```diff
@@ -23,6 +23,10 @@ extern "C" {
 #define LV_ARRAY_DEFAULT_CAPACITY  4
 #endif
 
+#ifndef LV_ARRAY_DEFAULT_SHRINK_RATIO
+#define LV_ARRAY_DEFAULT_SHRINK_RATIO 2
```

Comment:
I'm not sure if we shall define it as a ratio. Having and absolute number, e.g. `#define LV_ARRAY_MAX_FREE_CAPACITY` would be more predictable.

Reply:
> I'm not sure if we shall define it as a ratio. Having and absolute number, e.g. `#define LV_ARRAY_MAX_FREE_CAPACITY` would be more predictable.

I believe that defining the memory release size as an absolute value may not always be suitable. For instance, if it is set to 8, frequent resizing may still be necessary for large lists exceeding 256 elements, while small lists comprising less than 8 elements may no chance to be released

Reply:
> I think we need to pick the strategy to our use cases. Now we are using arrays for
> 
> * **events**: it's rare the you have >4 events for a widget, but there are many widgets. Removing events is quite rare too. Here immediate shrink would be a better choice.
> * **VGLite pending lists**: Usually small too (4, or 16) but they might change more frequently. cc @FASTSHIFT
> 
> In light of that (for small lists) an absolute shrink value (e.g. 4) seems like a better fit.

I think a ratio based posting rule would also be good for small lists, it would handle this more intelligently.

Reply:
This how it looks like when starting from 16 and keep reducing by 1.

Reduce is better at 12..9 but with smaller number ratio is better. Actually I'm convinced now that ratio is better in our case. Thank you for not giving up :slightly_smiling_face: 

| Size | Ratio (2) | Reduce (4) |
|------|-----------|------------|
| 16   | 16        | 16         |
| 15   | 16        | 16         |
| 14   | 16        | 16         |
| 13   | 16        | 16         |
| 12   | 16        | 12         |
| 11   | 16        | 12         |
| 10   | 16        | 12         |
| 9    | 16        | 12         |
| 8    | 8         | 8          |
| 7    | 8         | 8          |
| 6    | 8         | 8          |
| 5    | 8         | 8          |
| 4    | 4         | 4          |
| 3    | 4         | 4          |
| 2    | 2         | 4          |
| 1    | 1         | 4          |
| 0    | 0         | 0          |